runtime/runtime-params-estimator/src/lib.rs

```diff
@@ -1042,9 +1045,115 @@ fn storage_remove_ret_value_byte(ctx: &mut EstimatorContext) -> GasCost {
 }
 
 fn touching_trie_node(ctx: &mut EstimatorContext) -> GasCost {
-    // TODO: Actually compute it once our storage is complete.
-    // TODO: temporary value, as suggested by @nearmax, divisor is log_16(20000) ~ 3.57 ~ 7/2.
-    storage_read_base(ctx) * 2 / 7
+    let read = touching_trie_node_read(ctx);
+    let write = touching_trie_node_write(ctx);
+    return std::cmp::max(read, write);
+}
+
+fn touching_trie_node_read(ctx: &mut EstimatorContext) -> GasCost {
+    if let Some(cost) = ctx.cached.touching_trie_node_read.clone() {
+        return cost;
+    }
+    let warmup_iters = ctx.config.warmup_iters_per_block;
+    let measured_iters = ctx.config.iter_per_block;
+    let mut testbed = ctx.testbed();
+    let tb = testbed.transaction_builder();
+
+    // Number of bytes in the final key. Will create 2x that many nodes.
+    // Picked somewhat arbitrarily, balancing estimation time vs accuracy.
+    let final_key_len = 1000;
+
+    // Prepare a long chain in the trie
+    let signer = tb.random_account();
+    let key = "j".repeat(final_key_len);
+    let mut setup_block = Vec::new();
+    for key_len in 0..final_key_len {
+        setup_block.push(tb.account_insert_key(signer.clone(), &key.as_str()[..key_len], "0"));
+    }
+
+    let mut blocks = Vec::with_capacity(1 + 2 * warmup_iters + 2 * measured_iters);
+    blocks.push(setup_block);
+
+    blocks.extend(
+        iter::repeat_with(|| vec![tb.account_has_key(signer.clone(), &key.as_str()[0..1])])
+            .take(measured_iters + warmup_iters),
+    );
+    blocks.extend(
+        iter::repeat_with(|| vec![tb.account_has_key(signer.clone(), &key)])
+            .take(measured_iters + warmup_iters),
+    );
+
+    let results = &testbed.measure_blocks(blocks)[1..];
+    let (short_key_results, long_key_results) = results.split_at(measured_iters + warmup_iters);
+
+    let (cost_short_key, ext_cost_short_key) = aggregate_per_block_measurements(
+        &ctx.config,
+        1,
+        short_key_results[warmup_iters..].to_vec(),
+    );
+    let (cost_long_key, ext_cost_long_key) =
+        aggregate_per_block_measurements(&ctx.config, 1, long_key_results[warmup_iters..].to_vec());
+
+    let nodes_touched_delta = ext_cost_long_key[&ExtCosts::touching_trie_node]
+        - ext_cost_short_key[&ExtCosts::touching_trie_node];
```

Comment:
Well, it's not quite as simple as it sounds. While it is true that each additional key *produces* exactly two extra nodes, the exact number of *touched* nodes is trickier.

The numbers observed are:


|  | Key length  | Touched nodes |
| --- | --- | --- |
| Read | 1 byte | 6
| Write | 1 byte | 7
| Read | 1000 byte | 2001
| Write | 1000 byte | 2001


So the diffs to assert on would be 1995 and  1994. I can explain 1995 but 1994 still doesn't make sense to me.

The bottom line for me is, since these numbers are not trivial to understand, maybe we should not have these assertions. They might fail when a small implementation detail for the trie changes.


<details>
<summary>
expand only IF you are curious, here is my explanation for 1995.
</summary>

So the trie shape created looks a bit lik this:
```
       [Account root]
             |
[extension for a half-byte]
             |
         [Branch] --- [ Leaf ]       // j
             |
[extension for a half-byte]
             |
         [Branch] --- [ Leaf ]       // jj
             |
            ...
             |
[extension for a half-byte]
             |
         [Branch] --- [ Leaf ]       // j*999
             |
[extension for *three* half-bytes]
             |
         [ Leaf ]       // j*1000
```

The account root needs to be loaded in all cases, `X` nodes are touched for that.

Reading a one-byte-key touches 1 branch + 1 extension + 1 leaf + `X`.

The key with `n` bytes will lookup `(n-1)` branches + `(n-1)` extensions + 1 leaf + `X`. (It is `n-1` and not `n` because of the last extension being longer than the others.)

The difference between the two simplifies to `2*n-5`, thus for `n=1000` we get 1995.
(Also, we find that `X=3` but this is irrelevant for the formula.)

For writes, it should be the exact same story. But nope, somehow there is an additional lookup but only for the short key. Order doesn't matter here, it is always the short key that has this extra lookup. Maybe I will figure it out eventually. But the complexity alone is reason enough for me to question assertions for exact numbers.

</details>

Reply:
Yes that's reasonable. I've added an assertion on +/- 10 nodes with a comment that explains it.